modules/internal/ChapelHashing.chpl

```diff
@@ -25,14 +25,26 @@ module ChapelHashing {
 
   use ChapelBase;
 
-  proc chpl__defaultHashWrapper(x): int {
-    use Reflection;
-    if !canResolveMethod(x, "hash") then
-      compilerError("No hash function found for " + x.type:string);
+  interface Hashable {
+    proc Self.hash(): uint;
+  }
+
+  proc chpl__defaultHashWrapperCG(x: Hashable): int {
     const hash = x.hash();
     return (hash & max(int)): int;
   }
 
+  pragma "last resort"
+  proc chpl__defaultHashWrapper(x): int {
+    use Reflection;
+    if !canResolve("chpl__defaultHashWrapperCG", x) {
```

Comment:
I was expecting that we would add a primitive "Does it implement an interface". I think that doing so would address some of the weird compiler error behavior you are currently facing.